Feature: Add Subfinder Boefje

joeppeeters opened the pull request:
## Changes
I created a Boefje which wraps the Subfinder tool by projectdiscovery.io to discover subdomains. This might be relevant for your project. Let me know if this is something you would like to merge. If so, I might need to put some extra effort in checking all the boxes below.

## Issue ticket number and link
N/a

## Proof
Boefje tile with relevant image.
<img width="268" alt="Screenshot 2023-03-14 at 21 13 52" src="https://user-images.githubusercontent.com/3350917/225885375-e73bab29-ea7b-4457-96e9-8367279a5331.png">

Completed scan and normalizer
<img width="968" alt="Screenshot 2023-03-14 at 21 14 04" src="https://user-images.githubusercontent.com/3350917/225885383-45e6527a-c2c5-474b-bd26-377fa9c8405d.png">

Results as list of new Hostnames
<img width="961" alt="Screenshot 2023-03-14 at 21 14 45" src="https://user-images.githubusercontent.com/3350917/225885384-26280c7b-8157-4068-97b2-469286893915.png">

## Extra instructions for others

## Checklist for author(s):
- [x] All the commits in this PR are properly PGP-signed and verified;
- [ ] This PR comes from a `feature` or `hotfix` branch, in line with our git branching strategy;
- [x] This PR is "bite-sized" and only focuses on a single issue, problem, or feature;
- [ ] I am not reinventing the wheel: there is no high-quality library that already has this feature;
- [ ] I have changed the example `.env` files if I added, removed, or changed any config options, and I have informed others that they need to modify their `.env` files if required;
- [x] I have performed a self-review of my own code;
- [x] I have commented my code, particularly in hard-to-understand areas;
- [x] I have made corresponding changes to the documentation, if necessary;
- [ ] I have written unit, integration, and end-to-end tests for the change that I made;

If a non-trivial PR:
- [ ] This PR is part of a milestone and has appropriate labels;
- [ ] This PR is properly linked to the project board (either directly or via an issue);
- [x] I have added screenshots or some other proof that my code does what it is supposed to do;


```
## Checklist for functional reviewer(s):
- [ ] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [ ] I have checked out this branch, and successfully ran `make kat`;
- [ ] I have ran `make test-rf` and all end-to-end Robot Framework tests pass;
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended;
- [ ] I confirmed that there are no unintended functional regressions in this branch;

### What works:
* _bullet point + screenshot (if useful) per tested functionality_

### What doesn't work:
* _bullet point + screenshot (if useful) per tested functionality_

### Bug or feature?:
* _bullet point + screenshot (if useful) if it is unclear whether something is a bug or an intended feature._
```

```
## Checklist for code reviewer(s):
- [ ] The code passes the CI tests and linters;
- [ ] The code does not bypass authentication or security mechanisms;
- [ ] The code does not introduce any dependency on a library that has not been properly vetted;
- [ ] The code does not violate Model-View-Template and our other architectural principles;
- [ ] The code contains docstrings, comments, and documentation where needed;
- [ ] The code prioritizes readability over performance where appropriate;
- [ ] The code conforms to our agreed coding standards.
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/minvws/nl-kat-coordination?pullRequest=516) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/minvws/nl-kat-coordination?pullRequest=516) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/minvws/nl-kat-coordination?pullRequest=516) it.</sub>

olafz commented on boefjes/boefjes/plugins/pdio_subfinder/main.py:
```diff
@@ -0,0 +1,16 @@
+
+from boefjes.job_models import BoefjeMeta
+
+SSLSCAN_IMAGE = "projectdiscovery/subfinder:latest"
```
This was probably copy-pasted from another boefje? `SSLSCAN_IMAGE` => `SUBFINDER_IMAGE`?

underdarknl commented:
ah,wow. we'll have a look asap. :)

ammar92 commented on boefjes/boefjes/plugins/pdio_subfinder/cover.jpg:
Perhaps the cover image can be compressed?

ammar92 reviewed (commented):
Looks good! I'm currently testing this out and I only had a few small remarks 👍

ammar92 commented:
Seems to work 👍 It does however give a lot of false positives on some hosts. E.g. scanning tweakers.net hostname yields a lot of 301s which are seen as subdomains but in reality aren't

joeppeeters commented:
Thanks for your review! I'll address your remarks and the pre-commit stuff later this week.

> It does however give a lot of false positives on some hosts

Do you consider this a problem? Running it with `-active -ip` yields the following output:

`docker run projectdiscovery/subfinder:latest -json -active -ip -d tweakers.net`
  
```json
{"host":"true.tweakers.net","ip":"213.239.154.1","input":"tweakers.net","source":"anubis"}
{"host":"uploads.tweakers.net","ip":"213.239.154.49","input":"tweakers.net","source":"anubis"}
{"host":"mx2.tweakers.net","ip":"149.13.75.27","input":"tweakers.net","source":"anubis"}
{"host":"arethusa.tweakers.net","ip":"213.239.154.12","input":"tweakers.net","source":"anubis"}
{"host":"upload.tweakers.net","ip":"213.239.154.49","input":"tweakers.net","source":"anubis"}
{"host":"achelois.tweakers.net","ip":"213.239.154.10","input":"tweakers.net","source":"anubis"}
{"host":"adonis.tweakers.net","ip":"213.239.154.11","input":"tweakers.net","source":"anubis"}
{"host":"mx1.tweakers.net","ip":"130.117.54.106","input":"tweakers.net","source":"anubis"}
{"host":"irc.tweakers.net","ip":"31.22.80.151","input":"tweakers.net","source":"anubis"}
{"host":"mail.tweakers.net","ip":"213.239.154.11","input":"tweakers.net","source":"anubis"}
```

Does that look better? Would it make sense to return other OOIs too? Like an IPAddress to deal with duplicates?

ammar92 commented:
> Do you consider this a problem? Running it with `-active -ip` yields the following output:

I don't think it's a huge problem per se, because in a real world scenario you could have found a lot OOIs such as hostnames that get disconnected at some point. But giving too many of these could potentially slow down the application (e.g. due to (re)scanning of non-existing or invalid input OOIs)

> `docker run projectdiscovery/subfinder:latest -json -active -ip -d tweakers.net`
> 
> ```json
> {"host":"true.tweakers.net","ip":"213.239.154.1","input":"tweakers.net","source":"anubis"}
> {"host":"uploads.tweakers.net","ip":"213.239.154.49","input":"tweakers.net","source":"anubis"}
> {"host":"mx2.tweakers.net","ip":"149.13.75.27","input":"tweakers.net","source":"anubis"}
> {"host":"arethusa.tweakers.net","ip":"213.239.154.12","input":"tweakers.net","source":"anubis"}
> {"host":"upload.tweakers.net","ip":"213.239.154.49","input":"tweakers.net","source":"anubis"}
> {"host":"achelois.tweakers.net","ip":"213.239.154.10","input":"tweakers.net","source":"anubis"}
> {"host":"adonis.tweakers.net","ip":"213.239.154.11","input":"tweakers.net","source":"anubis"}
> {"host":"mx1.tweakers.net","ip":"130.117.54.106","input":"tweakers.net","source":"anubis"}
> {"host":"irc.tweakers.net","ip":"31.22.80.151","input":"tweakers.net","source":"anubis"}
> {"host":"mail.tweakers.net","ip":"213.239.154.11","input":"tweakers.net","source":"anubis"}
> ```
> 
> Does that look better? Would it make sense to return other OOIs too? Like an IPAddress to deal with duplicates?

This does look a lot better in my opinion and when manually testing these out I can clearly see a different response. 

Although you want to keep plugins simple, it would definitely make sense to return other OOIs too. In general KAT principles: the more information you could extract from the raw output, the better.

joeppeeters commented:
I addressed the comments and the pre-commit should now pass.

It now only returns `-active` domains, but I did not yield IP-addresses as the tool only resolves A records. It's probably best to use a proper DNS resolver here.

Let me know if anything else needs to change.

ammar92 reviewed (approved):
Looks good to me! Just one simple pre-commit issue needs to be fixed, but consider it approved 👍

joeppeeters commented:
Thanks for the review and fix 👍🏻
